_posts/2018-06-21-vision-for-rust-and-wasm.md

```diff
@@ -0,0 +1,255 @@
+`greet` function:
+
+```js
+import { greet } from "./hello_world";
+
+greet("World!");
+```
+
+<a href="/images/wasm-bindgen-architecture-current.png">
+  <img src="/images/wasm-bindgen-architecture-current.png" alt="wasm-bindgen architecture" style="float:right;width:50%;min-width:200px;max-width:400px;margin:1em"/>
+</a>
+
+How does `wasm-bindgen` work? Simplifying a bit, it is a procedural macro that
+takes in Rust source code annotated with `#[wasm_bindgen]` attributes,
+constructs an abstract syntax tree (AST), and then it emits two artifacts:
+
+1. Rust bindings that import JavaScript things and export Rust things.
+
+2. JavaScript bindings that expose a nice interface to Rust-exported things to
+   other JavaScript code and provide the Rust's desired imports.
+
+`wasm-bindgen`'s approach to JavaScript binding allows you to pay only for the
+imports that you use. Just because you imported the `window.alert` function, you
+don't end up with glue code for `window.document`.
+
+The big downside is that, right now, you always have to declare imports
+yourself. There are common imports for JavaScript functions and types and the
+Web platform APIs that will undoubtedly be repeated by many people many times
+over. Importing these by-hand is both boring and mechanical. We have a plan for
+fixing this, but you'll have to wait for a follow up blog post to learn more.
+
+<div style="clear: both"/>
+
+### wasm-pack
+
+[`wasm-pack` seeks to be a one-stop shop for building, optimizing, and
+publishing Rust-generated WebAssembly that you would like to interoperate with
+JavaScript, in the browser, or with Node.js.][wasm-pack] `wasm-pack` helps you
+build and publish Rust-generated WebAssembly to the npm registry to be used
+alongside any other JavaScript package in workflows that you already use, such
+as a bundler like [webpack][] or [greenkeeper][].
+
+[![wasm-pack cartoon](/images/wasm-pack-cartoon.png)](/images/wasm-pack-cartoon.png)
+
+*Drawing by Lin Clark in [Making WebAssembly better for Rust & for all
+languages][better-for-all]*
+
+The intention is that if you are a Rust developer and want to publish a crate
+compiled to wasm on npm, `wasm-pack` will
+
+1. compile the crate to WebAssembly with the `wasm32-unknown-unknown` target,
+2. run the `wasm-bindgen` CLI tool on the `.wasm` to generate its JavaScript
+   interface,
+3. run any other post-build tools such as `wasm-snip` and `wasm-opt`,
+4. collate any and all npm dependencies your crate and/or its JavaScript
+   bindings might have,
+5. and publish the resulting package on npm.
+
+All without you, the Rust developer, needing to have a JavaScript toolchain up
+and running.
+
+Right now, steps 1, 2, and 5 are in place, but you still need to have `npm`
+installed locally. There are also some more things planned for `wasm-pack`, and
+our story for orchestrating builds, dependencies, and publishing coming down the
+pipe, but you'll have to wait for the dedicated follow up blog post.
```

Comment:
Hmmmm since we keep saying this maybe we should have just one bit at the end saying we plan to follow up on each of these with a blog post detailing how you can help etc. This way we do it once to really get people excited, rather than oh this sounds interesting oh gotta wait for another blog post. Once is exciting too many feels draining.